Add NLog instrumentation for OpenTelemetry .NET Auto-Instrumentation

danifitz opened the pull request:
This commit implements a comprehensive NLog instrumentation that automatically bridges NLog logging events to OpenTelemetry without requiring code changes.

Features:
- Automatic target injection into NLog's target collection
- Complete log event bridging with level mapping
- Structured logging support with message templates
- Trace context integration
- Custom properties forwarding with filtering
- Comprehensive test coverage
- End-to-end test application
- Extensive documentation

The implementation follows the same patterns as the existing Log4Net instrumentation and supports NLog versions 4.0.0 through 6.*.*.

Configuration:
- OTEL_DOTNET_AUTO_LOGS_ENABLED=true
- OTEL_DOTNET_AUTO_LOGS_ENABLE_NLOG_BRIDGE=true
- OTEL_DOTNET_AUTO_LOGS_INCLUDE_FORMATTED_MESSAGE=true (optional)

Files added:
- src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/
- test/OpenTelemetry.AutoInstrumentation.Tests/NLogTests.cs
- test/test-applications/integrations/TestApplication.NLog/

Files modified:
- Configuration classes to support NLog bridge
- Public API files to include new integration class

## Why

https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/3938

Fixes #3938 

## What

This PR implements a comprehensive NLog instrumentation that automatically bridges NLog logging events to OpenTelemetry without requiring any code changes from users.

## Tests

- **Unit Tests**: 16 tests covering level mapping and edge cases
- **Integration Tests**: Complete end-to-end test application
- **Compatibility**: Verified on .NET 8.0 and .NET 9.0 with NLog 5.3.2

## Checklist

<!-- All items should be verified and marked as done.
     ~~strikethrough~~ if an item doesn't apply to the introduced changes. -->

- [ ] `CHANGELOG.md` is updated.
- [ ] Documentation is updated.
- [X] New features are covered by tests.

matt-hensley commented on test/test-applications/integrations/TestApplication.NLog/TestApplication.NLog.csproj:
```diff
@@ -0,0 +1,16 @@
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
+</Project> 
```
nit: missing new line

nrcventura reviewed (commented):
Thank you for submitting this PR. In general, I think these changes look good, but I want to give more time for those that worked on the other bridge implementation to review this PR too.

lachmatt commented on src/OpenTelemetry.AutoInstrumentation/Generated/net8.0/SourceGenerators/SourceGenerators.InstrumentationDefinitionsGenerator/InstrumentationDefinitions.g.cs:
```diff
@@ -104,6 +104,12 @@ private static NativeCallTargetDefinition[] GetDefinitionsArray()
             }
+
+            // NLog
+            if (logSettings.EnabledInstrumentations.Contains(LogInstrumentation.NLog))
```
Similar change seems to be missing from net462 file.

snakefoot commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/OpenTelemetryNLogTarget.cs:
```diff
@@ -0,0 +1,292 @@
+/// The target integrates with NLog's architecture by implementing the target pattern,
+/// allowing it to receive log events and forward them to OpenTelemetry for processing.
+/// </summary>
+internal class OpenTelemetryNLogTarget
```
NLog encourages the ability to configure the NLog Target from code or from NLog.config-file, instead of auto-instrumentation. Ofcourse good default values are encourages so the NLog Targets works out-of-the-box with no or little configuration.

When using `TargetWithContext` then one can use the NLog Layout abilities to enrich the OpenTelemetry-Logging-output with additional logeevent-properties. The NLog LoggingRules provides a lot of flexibility in filtering / routing NLog Logging output to the wanted target-destinations, but requires NLog targets.

danifitz replied:
I'm taking a week off to go on holiday - i wil look to refactor this when I get back, thanks

zacharycmontoya replied:
Hi folks, I just now took a look at this PR and [added a comment](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371#discussion_r2317206504) to request that we in fact go back to zero-code, so the original implementation of the OpenTelemetryNLogTarget looks like exactly what we want.

I apologize that I didn't comment sooner that zero-code is the direction we should take with this work. The reason for favoring this approach is that this project should not contain any other library dependencies except for the OTel SDK, so we should not be creating derived types that extend types from the NLog library.

Kielek commented on Directory.Packages.props:
```diff
@@ -6,6 +6,8 @@
     <PackageVersion Include="DotNet.ReproducibleBuilds" Version="1.2.25" />
     <PackageVersion Include="log4net" Version="3.1.0" />
     <PackageVersion Include="MinVer" Version="6.0.0" />
+    <PackageVersion Include="NLog" Version="5.3.2" />
+    <PackageVersion Include="NLog.Extensions.Logging" Version="5.3.11" />
```
It should be defined under `test/Directory.Packages.props`

Kielek replied:
log4net already in correct place: #4426

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/Integrations/LoggerIntegration.cs:
```diff
@@ -0,0 +1,73 @@
+maximumVersion: "6.*.*",
+integrationName: "NLog",
+type: InstrumentationType.Log)]
+public static class LoggerIntegration
```
@zacharycmontoya, could you please advice if it is the best integration point for this behavior?
As I remember DD supports NLog bridges/context injection.

Kielek commented on test/test-applications/integrations/TestApplication.NLogBridge/TestApplication.NLog.csproj:
Redundant file?

Kielek commented on test/test-applications/integrations/TestApplication.NLogBridge/TestApplication.NLogBridge.csproj:
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <ItemGroup>
+      <PackageReference Include="NLog" VersionOverride="$(LibraryVersion)" />
+      <PackageReference Include="NLog.Extensions.Logging" VersionOverride="$(LibraryVersion)" />
+      <PackageReference Include="Microsoft.Extensions.Logging"/>
+      <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" />
+      <PackageReference Include="System.Diagnostics.DiagnosticSource" />
```
I am not fully familiar with NLog details. Is it necessary to bring all these dependencies to make it working? Can this list be reduced?

danifitz replied:
Addressed in [799ef71](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/799ef710f2799637667a61031f2dc6d48b4e149e)

Kielek commented on build/LibraryVersions.g.cs:
```diff
@@ -68,6 +68,13 @@ public static partial class LibraryVersion
+            "TestApplication.NLogBridge",
+            [
+                new("4.7.15"),
+                new("5.3.2"),
```
I think that this list should contain also 6.0.0+

Kielek reviewed (commented):
@danifitz, @snakefoot - I like the idea to introduce support for NLog here. If some/all cases cane be done by NLog directly, it is great (maybe context correlation can be done, without any changes here and we only need to document it/enable by default?) 
The goal of the support should be also possibility to automatically also sent data through OTLP without any changes in the configuration/code.

snakefoot commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+
+    public bool UseHttp { get; set; } = true;
+
+    public string? ServiceName { get; set; }
```
Change from `string?` to NLog `Layout?`

snakefoot commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+        var factory = _getLoggerFactory;
+        if (factory is null)
+        {
+            return new object();
```
This looks weird. Why not allow `object?` and just discards LogEvents when `GetOrCreateLogger` returns `null` ?

snakefoot commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+        var current = Activity.Current;
+
+        // Emit using internal helpers via reflection delegate
+        var renderedMessage = logEvent.FormattedMessage;
```
Instead change to:
```c#
var renderedMessage = RenderLogEvent(Layout, logEvent);
```

snakefoot commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+
+        // Emit using internal helpers via reflection delegate
+        var renderedMessage = logEvent.FormattedMessage;
+        var args = IncludeEventParameters && logEvent.Parameters is object[] p ? p : null;
```
Think you should only `IncludeEventParameters` when `logEvent.HasProperties == false`

snakefoot commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+
+        // Build properties from event properties and context
+        var properties = new List<KeyValuePair<string, object?>>();
+        if (IncludeEventProperties && logEvent.HasProperties && logEvent.Properties is not null)
```
`logEvent.Properties` is never `null`

snakefoot commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+        // Scope properties can be added via explicit <attribute> entries or NLog's contexts (GDC/MDLC)
+        foreach (var attribute in Attributes)
+        {
+            var value = attribute.Layout?.Render(logEvent);
```
instead use the following
```c#
var value = RenderLogEvent(attribute.Layout, logEvent);
```

danifitz replied:
resolved in [4401757](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/440175748c7fc3d7ee52f84af92a7d813686ed9d)

snakefoot commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+
+        loggerProviderBuilder = loggerProviderBuilder.AddOtlpExporter(options =>
+        {
+            var endpoint = Endpoint;
```
When using NLog Layout, then do like this:
```c#
var endpoint = RenderLogEvent(Endpoint, LogEventInfo.CreateNullEvent());
```

danifitz replied:
resolved in [4401757](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/440175748c7fc3d7ee52f84af92a7d813686ed9d)

snakefoot commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+    public string? ServiceName { get; set; }
+
+    [ArrayParameter(typeof(TargetPropertyWithContext), "attribute")]
+    public IList<TargetPropertyWithContext> Attributes { get; } = new List<TargetPropertyWithContext>();
```
Any reason for defining new `Attributes`, instead of just relying on existing from base-class ? (Called `Properties`)

danifitz replied:
resolved in [4401757](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/440175748c7fc3d7ee52f84af92a7d813686ed9d)

danifitz commented:
Thank you @snakefoot for your speedy and detailed feedback, so invaluable. I've made some edits (all in one commit) [4401757](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/440175748c7fc3d7ee52f84af92a7d813686ed9d) to address your feedback.

# Summary of Changes
Some Architectural Improvements
* AsyncWrapper Compatibility: Replaced Activity.Current with Layout-based trace context resolution using TraceIdLayout and SpanIdLayout properties
* Performance Optimization: Added HasProperties check to avoid unnecessary dictionary allocation
* Standards Compliance: Removed custom Attributes property in favor of base class ContextProperties
* Layout Consistency: All rendering operations now use RenderLogEvent() pattern
* Property Logic Refinement: Enhanced parameter vs. properties handling logic
* Memory Efficiency: Eliminated dummy object creation in favor of null returns

danifitz commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/OpenTelemetryNLogConverter.cs:
```diff
@@ -0,0 +1,199 @@
+        try
+        {
+            var properties = loggingEvent.GetProperties();
+            return properties == null ? null : GetFilteredProperties(properties);
```
resolved in [4401757](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/440175748c7fc3d7ee52f84af92a7d813686ed9d)

danifitz commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+
+    public bool IncludeFormattedMessage { get; set; } = true;
+
+    public new bool IncludeEventProperties { get; set; } = true;
```
resolved in [4401757](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/440175748c7fc3d7ee52f84af92a7d813686ed9d)

danifitz commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,211 @@
+    [RequiredParameter]
+    public string? Endpoint { get; set; }
+
+    public string? Headers { get; set; }
```
resolved in [4401757](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/440175748c7fc3d7ee52f84af92a7d813686ed9d)

danifitz commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,242 @@
+            {
+                var traceId = ActivityTraceId.CreateFromString(traceIdString);
+                var spanId = ActivitySpanId.CreateFromString(spanIdString);
+                var activityContext = new ActivityContext(traceId, spanId, ActivityTraceFlags.Recorded);
```
My observation is that there are plenty of enterprise companies with older versions of .NET/Nlog who want to use OpenTelemetry. However I am fine with using typed Layout for now and see if we get any issues raised on the repo looking for support for older NLog version.

danifitz replied:
Resolved in [0fcf644](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/0fcf644d2da62301dfcdef13e1b930c94d6960e7)

I tried to take a hybrid approach:

# NLog Hybrid Layout Implementation

## Version Detection
- **Runtime Check**: Detects NLog version using `typeof(Layout).Assembly.GetName().Version`
- **Threshold**: NLog 5.3.4+ supports `Layout<T>` typed layouts
- **Fail-Safe**: Defaults to string parsing if version detection fails

## Backward Compatibility
- **Public API**: Unchanged `Layout? TraceIdLayout` and `Layout? SpanIdLayout` properties
- **Version Range**: Supports NLog 4.0.0 through 6.*.*

## Fallback Strategy

### NLog 5.3.4+ (Modern)
```csharp
// Direct typed rendering - no parsing overhead
traceId = _typedTraceIdLayout.Render(logEvent);
spanId = _typedSpanIdLayout.Render(logEvent);
```

### NLog 4.0.0 - 5.3.3 (Legacy)
```csharp
// String rendering + parsing with error handling
var traceIdString = RenderLogEvent(TraceIdLayout, logEvent);
traceId = ActivityTraceId.CreateFromString(traceIdString);
```

Let me know your thoughts!

snakefoot replied:
The optimization with [type Layout](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371#discussion_r2288804643) comes from using `Layout<System.Diagnostics.ActivityTraceId?>.FromMethod(...)`

By assigning `_typedTraceIdLayout = "${activity:property=TraceId}"` then it will attempt to parse from string (and will probably fail)

zacharycmontoya commented on test/IntegrationTests/NLogBridgeTests.cs:
```diff
@@ -0,0 +1,189 @@
+            VerifyTraceContext(logRecord) &&
+            logRecord is { SeverityText: "Info", SeverityNumber: SeverityNumber.Info } &&
+            VerifyAttributes(logRecord) &&
+            logRecord.Attributes.Count == 4,
```
Using NLog directly has 4 log record attributes vs 3 log record attributes when using ILogger with the NLog provider. What's different in the sets of attribute?

danifitz replied:
The custom attribute `"test_key": "test_value"` is only present when using the NLog Bridge directly but gets filtered out when using the ILogger Bridge. 

This suggests:
NLog Bridge: Preserves all NLog-specific properties and custom attributes
ILogger Bridge: Filters out some NLog-specific attributes, keeping only the structured logging parameters

This makes sense because:
Direct NLog usage has full access to NLog's property system

ILogger usage goes through Microsoft's logging abstraction, which may not preserve all NLog-specific metadata
The test correctly validates that the two different logging paths (direct NLog vs ILogger provider) produce different attribute sets, which is expected behavior

zacharycmontoya commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/OpenTelemetryLogHelpers.cs:
```diff
@@ -0,0 +1,320 @@
+        var attributesExpression = BuildLogRecordAttributes(logRecordAttributesListType, exception, properties, args, renderedMessage);
+
+        // Get the EmitLogRecord method from the logger
+        var emitLogRecordMethod = loggerType.GetMethod("EmitLogRecord", new[] { logRecordDataType, logRecordAttributesListType })!;
```
The method in the SDK is [EmitLog](https://github.com/open-telemetry/opentelemetry-dotnet/blob/2a9c406b6da7d10dc9dfbee1f6528e2c85e4e075/src/OpenTelemetry/Logs/LoggerSdk.cs#L27). Is this working correctly?

danifitz replied:
Fixed [665fb09](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/665fb09957bbbe235dd94395d62687b3bf48be33)

zacharycmontoya commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/OpenTelemetryNLogConverter.cs:
```diff
@@ -0,0 +1,199 @@
+    public string Name { get; set; } = nameof(OpenTelemetryNLogConverter);
+
+    [DuckReverseMethod(ParameterTypeNames = new[] { "NLog.LogEventInfo, NLog" })]
+    public void WriteLogEvent(ILoggingEvent loggingEvent)
```
When is this WriteLogEvent being called? It seems like all of the calls go through OpenTelemetryTarget unless I'm missing something.

zacharycmontoya commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
In order to accept this pull request, I'm requesting a couple of changes:
- The `OpenTelemetryTarget` should be moved to the main `OpenTelemetry.AutoInstrumentation` project
- The `OpenTelemtryTarget` should not add any TypeReferences/AssemblyReferences to the NLog library and instead use our duck typing to create the runtime proxy types.

In this file, the reverse proxy code would like like this:
```csharp
[DuckReverseMethod(ParameterTypeNames = new[] { "NLog.LogEventInfo, NLog" })]
public void Write(ILoggingEvent? logEvent)
{
    // ...
}
```

Then in the `src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/AutoInjection/NLogAutoInjector.cs` file you would create a new `OpenTelemetryTarget` but wrap it in a reverse proxy type when registering it with NLog like:
```csharp
var otelTarget = new OpenTelemetryTarget();
var targetType = Type.GetType("NLog.Targets.TargetWithContext, NLog", false);
var targetProxy = otelTarget.DuckImplement(targetType);
addTargetMethod?.Invoke(configuration, new object?[] { "otlp", targetProxy });

var rule = Activator.CreateInstance(loggingRuleType!, new object?[] { "*", traceLevel, targetProxy });
```


These changes make the integration quite limiting since the user cannot configure them with XML configuration any more, but this ensures that we don't introduce any assembly loading issues due to assembly references from our assembly.

zacharycmontoya replied:
Looks good, thanks!

zacharycmontoya commented on src/OpenTelemetry.AutoInstrumentation.NLogTarget/OpenTelemetryTarget.cs:
```diff
@@ -0,0 +1,316 @@
+
+        loggerProviderBuilder = loggerProviderBuilder.AddOtlpExporter(options =>
+        {
+            var endpoint = RenderLogEvent(Endpoint, LogEventInfo.CreateNullEvent());
```
With my recommendation that we do not have programmatic configuration, we can rely solely on OTEL environment variables for the OTLP exporter settings, so I think we can remove the calls to `RenderLogEvent` here

danifitz commented on test/OpenTelemetry.AutoInstrumentation.Tests/NLogTests.cs:
```diff
@@ -0,0 +1,142 @@
+    [InlineData(2, 9)]   // Info (2) -> Should be Info (9)
+    [InlineData(3, 13)]  // Warn (3) -> Should be Warn (13)
+    [InlineData(4, 17)]  // Error (4) -> Should be Error (17)
+    public void CustomLevelsBetweenStandardLevels_AreMappedCorrectly(int nlogOrdinal, int expectedSeverity)
```
fixed in [f6cd7d5](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/f6cd7d5ad283b58f58568ecf99f07a44e8d268b4)

Kielek reviewed (commented):
Couple general comments,
* Please add entry to the CHANGELOG.md
* Please udpate config.md file
* Could you please test your solution with NLog 6+? I think that we should bring this support together with 5.*.

danifitz commented:
@Kielek I made those changes in [b078f85](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/b078f856c3c960f41ad6035ed9688649d972f7a0)

danifitz commented:
@Kielek @snakefoot @matt-hensley @zacharycmontoya I believe all comments/feedback resolved.

zacharycmontoya reviewed (approved):
LGTM. Thank you @danifitz for this contribution and being receptive to all the feedback along the way 🙏🏼

danifitz commented:
Thanks for cleaning this up @Kielek - any further action needed from me?

Kielek commented:
@danifitz, still reviewing. In general looks good, for applied small changes, it was easier to make commits directly than writing a comments. I hope you are fine with this.

danifitz commented:
> @danifitz, still reviewing. In general looks good, for applied small changes, it was easier to make commits directly than writing a comments. I hope you are fine with this.

I'm so fine with that :)

lachmatt commented on test/IntegrationTests/NLogBridgeTests.cs:
```diff
@@ -0,0 +1,189 @@
+            Arguments = "--api nlog"
+        });
+
+        var regex = new Regex(@"INFO  TestApplication\.NLogBridge\.Program - Hello, world at \d{2}\:\d{2}\! span_id=[a-f0-9]{16} trace_id=[a-f0-9]{32} trace_flags=01");
```
```suggestion
        var regex = new Regex(@"INFO TestApplication\.NLogBridge\.Program - Hello, world at \d{2}\:\d{2}\! span_id=[a-f0-9]{16} trace_id=[a-f0-9]{32} trace_flags=01");
```

lachmatt commented on test/IntegrationTests/NLogBridgeTests.cs:
```diff
@@ -0,0 +1,189 @@
+
+    private static void AssertStandardOutputExpectations(string standardOutput)
+    {
+        Assert.Contains("INFO  TestApplication.NLogBridge.Program - Hello, world at", standardOutput);
```
```suggestion
        Assert.Contains("INFO TestApplication.NLogBridge.Program - Hello, world at", standardOutput);
```

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/Integrations/LoggerIntegration.cs:
```diff
@@ -0,0 +1,109 @@
+            NLogAutoInjector.EnsureConfigured();
+
+            // Inject trace context into NLog GlobalDiagnosticsContext for current destination outputs
+            TrySetTraceContext(Activity.Current);
```
I think that this method should be called always, even if the bridge is not enabled.
Bridge means that we are sending NLog logs through OTel infrastructure. Typically OTLP.

You need to set context also when you are injecting in to the typical NLog sinks.

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/Integrations/LoggerIntegration.cs:
```diff
@@ -0,0 +1,109 @@
+/// The integration targets NLog.Logger.Log method which is the core method called
+/// for all logging operations, allowing us to capture events without modifying configuration.
+/// </summary>
+[InstrumentMethod(
+assemblyName: "NLog",
+typeName: "NLog.Logger",
+methodName: "Log",
+returnTypeName: ClrNames.Void,
+parameterTypeNames: new[] { "NLog.LogEventInfo" },
```
I think that you need to revisit the way how you injecting bytecode.

[`Logger.log(LogEventInfo)`](https://github.com/NLog/NLog/blob/809ef9cd79e76c6c53c6cccbd5778bd48579fdd0/src/NLog/Logger.cs#L268-L279) is not covering all cases when you are trying to log.

Even when I have checked the testing application calling this method, it is failing.

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/OpenTelemetryNLogConverter.cs:
```diff
@@ -0,0 +1,199 @@
+/// <summary>
+/// Converts NLog LogEventInfo into OpenTelemetry LogRecords.
+/// </summary>
+internal class OpenTelemetryNLogConverter
```
I do not found any usage of this class except the tests.

Kielek reviewed (changes requested):
Hi, I have spend some time trying to fix.

The infrastructure + documentation now looks good to me. The problematic part is that basically is not working to me, any of the implemented tests nor manual scenarios.

Please let me know if the comments are meaningful, if no I will try to provide more information.

danifitz commented:
I'm a bit busy over the next few days but will try and find some time to look into these comments next week or the weekend

ysolomchenko commented:
@danifitz, please add support for file-based configuration.

danifitz commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/Bridge/OpenTelemetryNLogConverter.cs:
```diff
@@ -144,38 +130,65 @@ internal static int MapLogLevel(int levelOrdinal)
         {
-            var properties = loggingEvent.Properties;
-            if (properties == null)
+            var gdcProperties = GlobalDiagnosticsContextGetter.Value?.Invoke();
```
That's fine with me :) - removed in [03f2c8b](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/03f2c8b0c9e9c38ffbdc625810a25516e66293e0)

danifitz commented:
Hi @Kielek the [failing CI test](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/actions/runs/19922305731/job/57115537445?pr=4371) seems to be unrelated to the NLog changes:

```
<Message>Assert.InRange() Failure: Value not in range
Range:  (50 - 70)
Actual: 0</Message>
          <StackTrace>   at IntegrationTests.SelectiveSamplerTests.ExportThreadSamples() in /home/runner/work/opentelemetry-dotnet-instrumentation/opentelemetry-dotnet-instrumentation/test/IntegrationTests/SelectiveSamplerTests.cs:line 51
   at System.RuntimeMethodHandle.InvokeMethod(ObjectHandleOnStack target, Void** arguments, ObjectHandleOnStack sig, BOOL isConstructor, ObjectHandleOnStack result)
   at System.RuntimeMethodHandle.InvokeMethod(ObjectHandleOnStack target, Void** arguments, ObjectHandleOnStack sig, BOOL isConstructor, ObjectHandleOnStack result)
   at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)</StackTrace>
```

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/TraceContextInjection/Integrations/WriteLogEventToTargetsIntegration.cs:
```diff
@@ -0,0 +1,41 @@
+    maximumVersion: "6.*.*",
+    integrationName: "NLog",
+    type: InstrumentationType.Log)]
+public static class WriteLogEventToTargetsIntegration
```
I think that this integration cover only part of the events going through NLog,

There are three `WriteLogEventToTargets` methods:

* `private void WriteToTargets(LogLevel level, Exception? ex, IFormatProvider? formatProvider, string message, object?[]? args)` - covered, it goes through 2 parameter methos
* `private void WriteLogEventToTargets(LogEventInfo logEvent, ITargetWithFilterChain targetsForLevel)` - covered
* `private void WriteLogEventToTargets(Type wrapperType, LogEventInfo logEvent, ITargetWithFilterChain targetsForLevel)` - not covered

The missing part can be reached by public api: `public void Log(Type wrapperType, LogEventInfo logEvent)`

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/TraceContextInjection/Integrations/WriteToTargetsIntegration.cs:
```diff
@@ -0,0 +1,46 @@
+    maximumVersion: "5.*.*",
+    integrationName: "NLog",
+    type: InstrumentationType.Log)]
+public static class WriteToTargetsIntegration
```
Similar to what I have posted for 6.0.0
Two independent methods needs to be covered
* `private void WriteToTargets(Type wrapperType, [NotNull] LogEventInfo logEvent, [NotNull] ITargetWithFilterChain targetsForLevel)` - not covered
* `private void WriteToTargets([NotNull] LogEventInfo logEvent, [NotNull] ITargetWithFilterChain targetsForLevel)` - covered

Kielek commented on src/OpenTelemetry.AutoInstrumentation/Instrumentations/NLog/TraceContextInjection/Integrations/WriteToTargetsLegacyIntegration.cs:
```diff
@@ -0,0 +1,46 @@
+    maximumVersion: "5.*.*",
+    integrationName: "NLog",
+    type: InstrumentationType.Log)]
+public static class WriteToTargetsLegacyIntegration
```
Similar to what I have posted for 6.0.0
Two independent methods needs to be covered
* `private void WriteToTargets(Type wrapperType, [NotNull] LogEventInfo logEvent, [NotNull] TargetWithFilterChain targetsForLevel)` - not covered
* `private void WriteToTargets([NotNull] LogEventInfo logEvent, [NotNull] TargetWithFilterChain targetsForLevel)` - covered

danifitz replied:
Fixed in [8a4483b](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4371/commits/8a4483b03e485cbd9dff223f09a9d3d71982bdbc)

Kielek commented on test/test-applications/integrations/TestApplication.NLogBridge/NLogLogger.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace TestApplication.NLogBridge;
+
+internal class NLogLogger : Microsoft.Extensions.Logging.ILogger
```
I would consider using this package https://github.com/NLog/NLog.Extensions.Logging
It is common ILogger provider.
log4net implements it in our repository, because similar package was not available at the moment of writing.
